### ISSUE TITLE
Training does not converge

### ISSUE THREAD

**Issue by Nico321** [shivam13juna/Quick_Image_Matting_remove.bg_equivalent#3]
I'm also looking for an image matting model like remove.bg while using it offline.
I tried training the segmentation and segmentation plus model. But it looks like the model is not improving while training. 

It outputs a complete white mask for all validation images like this one (result after 17 epochs):
![image](https://user-images.githubusercontent.com/6123294/97312773-261b8280-1866-11eb-9005-56df4c799b4d.png)

did you use specific learning rates or other hyperparameters other than the defaults provided in the train.py script? The loss only does minor improvements:
![image](https://user-images.githubusercontent.com/6123294/97312963-624ee300-1866-11eb-84f5-bed193358367.png)

same goes for the accuracy:
![image](https://user-images.githubusercontent.com/6123294/97313088-89a5b000-1866-11eb-9e73-b563b59c2d1c.png)

I could not try the trimap and matting models, because i do not have the extra masks. Would you mind sharing the masks or your trained models?

This is my training log:
```
epoch       main/loss   validation/main/loss  main/accuracy  validation/main/accuracy  lr          elapsed_time
0           0.691012                          0.633465                                 0.0001      40.0681       
0           0.685441                          0.659605                                 0.0001      78.5899       
0           0.680728                          0.655473                                 0.0001      117.196       
0           0.676663                          0.65249                                  0.0001      155.824       
0           0.674503                          0.641358                                 0.0001      194.202       
0           0.666601                          0.665989                                 0.0001      232.795       
0           0.666102                          0.649114                                 0.0001      271.254       
1           0.661145    0.665867              0.656182       0.641924                  0.0001      319.487       
1           0.657644                          0.656958                                 0.0001      358.066       
1           0.662633                          0.632791                                 0.0001      396.356       
1           0.655145                          0.649994                                 0.0001      434.874       
1           0.655596                          0.64355                                  0.0001      473.18        
1           0.651644                          0.649668                                 0.0001      511.46        
1           0.650833                          0.648181                                 0.0001      549.944       
2           0.64119     0.652668              0.666029       0.641924                  0.0001      594.518       
2           0.653462                          0.638094                                 0.0001      633.05        
2           0.641974                          0.659574                                 0.0001      671.617       
2           0.651572                          0.640113                                 0.0001      710.071       
2           0.643497                          0.654501                                 0.0001      748.618       
2           0.653847                          0.635408                                 0.0001      786.985       
2           0.642205                          0.655749                                 0.0001      825.493       
3           0.651805    0.650062              0.63872        0.641924                  0.0001      870.212       
3           0.646955                          0.647122                                 0.0001      908.858       
3           0.645811                          0.648793                                 0.0001      947.592       
3           0.643045                          0.653305                                 0.0001      986.167       
3           0.644804                          0.650208                                 0.0001      1024.95       
3           0.644504                          0.650382                                 0.0001      1063.67       
3           0.650888                          0.640078                                 0.0001      1102.22       
4           0.637512    0.649887              0.661703       0.641924                  0.0001      1146.78       
4           0.638354                          0.660014                                 0.0001      1185.31       
4           0.637932                          0.660207                                 0.0001      1223.75       
4           0.647008                          0.646334                                 0.0001      1262.18       
4           0.652052                          0.638102                                 0.0001      1300.83       
4           0.641764                          0.654494                                 0.0001      1339.31       
4           0.65452                           0.634254                                 0.0001      1377.76       
5           0.644559    0.649709              0.649931       0.641924                  0.0001      1428.58       
5           0.642224                          0.653626                                 0.0001      1467.04       
5           0.63227                           0.669191                                 0.0001      1505.51       
5           0.655237                          0.633206                                 0.0001      1543.97       
5           0.649148                          0.642573                                 0.0001      1582.46       
5           0.643584                          0.651266                                 0.0001      1620.94       
5           0.644439                          0.649845                                 0.0001      1659.44       
5           0.642787                          0.652324                                 0.0001      1697.9        
6           0.653855    0.649584              0.63501        0.641924                  0.0001      1742.45       
6           0.642423                          0.652957                                 0.0001      1780.99       
6           0.654993                          0.632995                                 0.0001      1819.52       
6           0.634141                          0.666174                                 0.0001      1858.03       
6           0.644173                          0.65                                     0.0001      1896.52       
6           0.641851                          0.653575                                 0.0001      1934.95       
6           0.64701                           0.64546                                  0.0001      1973.35       
7           0.637359    0.649459              0.660642       0.641924                  0.0001      2018.02       
7           0.638244                          0.659081                                 0.0001      2056.6        
7           0.645832                          0.647369                                 0.0001      2095.16       
7           0.641888                          0.653224                                 0.0001      2133.74       
7           0.651324                          0.638668                                 0.0001      2172.15       
7           0.651171                          0.638648                                 0.0001      2210.74       
7           0.642852                          0.651656                                 0.0001      2249.25       
8           0.639241    0.649246              0.657483       0.641924                  0.0001      2293.91       
8           0.641566                          0.653738                                 0.0001      2332.43       
8           0.64517                           0.647977                                 0.0001      2370.88       
8           0.64318                           0.650846                                 0.0001      2409.29       
8           0.647409                          0.644351                                 0.0001      2447.8        
8           0.638221                          0.658767                                 0.0001      2486.33       
8           0.649381                          0.641389                                 0.0001      2524.77       
9           0.649235    0.649107              0.641385       0.641924                  0.0001      2569.31       
9           0.641149                          0.65405                                  0.0001      2607.89       
9           0.640695                          0.654559                                 0.0001      2646.36       
9           0.638619                          0.657964                                 0.0001      2684.88       
9           0.639721                          0.656081                                 0.0001      2723.42       
9           0.641949                          0.652542                                 0.0001      2761.89       
9           0.643446                          0.650116                                 0.0001      2800.34       
10          0.649837    0.648936              0.640237       0.641924                  0.0001      2850.93       
10          0.639756                          0.655904                                 0.0001      2889.39       
10          0.646186                          0.645804                                 0.0001      2927.9        
10          0.648072                          0.642713                                 0.0001      2966.27       
10          0.643192                          0.650374                                 0.0001      3004.87       
10          0.65324                           0.634645                                 0.0001      3043.33       
10          0.642307                          0.65194                                  0.0001      3081.87       
10          0.637466                          0.659434                                 0.0001      3120.25       
11          0.644469    0.648757              0.648379       0.641924                  0.0001      3164.78       
11          0.637836                          0.658462                                 0.0001      3203.26       
11          0.650083                          0.639538                                 0.0001      3241.72       
11          0.641305                          0.65312                                  0.0001      3280.34       
11          0.642995                          0.650332                                 0.0001      3318.76       
11          0.652189                          0.636065                                 0.0001      3357.3        
11          0.641199                          0.653118                                 0.0001      3395.82       
12          0.642645    0.648561              0.650843       0.641924                  0.0001      3440.45       
12          0.639209                          0.656186                                 0.0001      3479.13       
12          0.652013                          0.636338                                 0.0001      3517.69       
12          0.637697                          0.658439                                 0.0001      3556.25       
12          0.635728                          0.661191                                 0.0001      3594.81       
12          0.656665                          0.628987                                 0.0001      3633.47       
12          0.6388                            0.656285                                 0.0001      3672.16       
13          0.643627    0.648361              0.648905       0.641924                  0.0001      3716.79       
13          0.645308                          0.646206                                 0.0001      3755.39       
13          0.641705                          0.652198                                 0.0001      3793.83       
13          0.647921                          0.642048                                 0.0001      3832.23       
13          0.641483                          0.652195                                 0.0001      3870.84       
13          0.635713                          0.660907                                 0.0001      3909.55       
13          0.640355                          0.65383                                  0.0001      3948.14       
14          0.649499    0.648173              0.639675       0.641924                  0.0001      3992.95       
14          0.65215                           0.635025                                 0.0001      4031.45       
14          0.648154                          0.641409                                 0.0001      4069.95       
14          0.649873                          0.638588                                 0.0001      4108.45       
14          0.636246                          0.660558                                 0.0001      4147          
14          0.635391                          0.661299                                 0.0001      4185.51       
14          0.640017                          0.653885                                 0.0001      4223.97       
15          0.634679    0.648134              0.662151       0.641924                  0.0001      4274.49       
15          0.641345                          0.651712                                 0.0001      4312.88       
15          0.636243                          0.659362                                 0.0001      4351.44       
15          0.631459                          0.66675                                  0.0001      4389.91       
15          0.652099                          0.63581                                  0.0001      4428.46       
15          0.655058                          0.630566                                 0.0001      4467.14       
15          0.649286                          0.639026                                 0.0001      4505.87       
15          0.638133                          0.656658                                 0.0001      4544.43       
16          0.639357    0.647774              0.654638       0.641924                  0.0001      4589.02       
16          0.634255                          0.662202                                 0.0001      4627.59       
16          0.649478                          0.639165                                 0.0001      4667.01       
16          0.656962                          0.627085                                 0.0001      4705.57       
16          0.634792                          0.661793                                 0.0001      4743.96       
16          0.643555                          0.647843                                 0.0001      4783.89       
16          0.638741                          0.655501                                 0.0001      4822.28       
17          0.636844    0.647647              0.658051       0.641924                  0.0001      4866.83       
17          0.64211                           0.65012                                  0.0001      4905.57       
17          0.638003                          0.656115                                 0.0001      4944.23       
17          0.641151                          0.651186                                 0.0001      4982.78       
17          0.643968                          0.647199                                 0.0001      5021.4        
17          0.650125                          0.637638                                 0.0001      5060.02       
17          0.643678                          0.647045                                 0.0001      5098.58       
18          0.642073    0.647379              0.649535       0.641924                  0.0001      5143.2        
18          0.634822                          0.660864                                 0.0001      5181.67       
18          0.641939                          0.649834                                 0.0001      5220.29       
18          0.638194                          0.655524                                 0.0001      5258.74       
18          0.632349                          0.664204                                 0.0001      5297.44 
``` 

And the train args:
```JSON
{
    "config": "config.json",
    "batchsize": 1,
    "max_iteration": 30000,
    "gpus": [
        0
    ],
    "lr": 0.0001,
    "momentum": 0.99,
    "weight_decay": 0.0005,
    "out": "result",
    "resume": "",
    "mode": "seg+",
    "pretrained_fcn8s": null,
    "pretrained_n_input_ch": 3,
    "pretrained_n_output_ch": 21,
    "mat_scale": 4
}
```

**Comment by shivam13juna** [comment]
Hello Nico, 

When I developed this code, I was given the mask by default. My client had these masks already and didn't tell me how he generated it. But I'm given to understand the generation of these masks isn't very difficult. I've also listed a method for generating mask in the repo, did you try that? 

The thing is, most people out there that want to do image matting have masks, they don't know how to extract people from the image using the mask with very high precision, that's where my repo is intended to help. 

Please let me know how you figure your problem, it'll help many people who're trying to replicate the results. Also in folder example 1, 2, and 3. There are sample masks( I was given, not generated myself). So you can have an idea of the precision of the mask required. 

Good Luck!

**Comment by Nico321** [comment]
Hi,

thanks for you quick reply.

Yes i tried training the segmentation model from your mask_gen folder using the `mask_gen/scripts/train.py` script. If I understood correct i could then use the model to generate the masks. But the model training seems to always predict a full mask and does not train very well. 

Currently I do have quite satisfying results using a [U^2-Net (U square net)](https://github.com/NathanUA/U-2-Net) which i fine tuned on the same dataset you were using in mask_gen, while increasing the training image size from 320x320 to 640x640. This model alone give already very precise masks like this one:
![image](https://user-images.githubusercontent.com/6123294/97670387-635e5b00-1a86-11eb-89dc-5992dad94148.png)
which i could further improve using image matting. But in some cases its not on the remove.bg - level of quality. E.g. it sometimes cuts a small piece of the hair or shoulder. But I've used a different matting algorithm then the one you proposed, but I've also splitted the image into 16 slices. I might give your matting algorithm a try and see if it helps.

**Comment by shivam13juna** [comment]
The mask you've attached seems like a very good quality mask. And please try my method, it won't take more than 15 minutes to set up. Please let me know the results I'm very curious. 

Also, the purpose of dividing the image into multiple slices is to do it in real-time, for example, if you have this algorithm on a website and a user uploads an image, he probably wouldn't want to wait for more than a minute right?  Other than time reasons, you can probably guess dividing the image has no other benefits. 

Also please attach an image generated from my algorithm and the one you tried, it'll again help everyone. Good Luck Nico!

**Comment by Nico321** [comment]
Ah okay, I assumed it would also improve the quality. Performance wise its okay to do it as one. We use this model in an REST API for a mobile app. But we do it on decent GPU and a few seconds processing time is accepted by our customer.

Actually the example was one of the better masks. The original image was on a white background. With noisy backgrounds the masks get worse like for this image I've found on google:
![image](https://user-images.githubusercontent.com/6123294/97676784-73c80300-1a91-11eb-865c-a765b977eaff.png)
results in this mask:
![image](https://user-images.githubusercontent.com/6123294/97676818-82161f00-1a91-11eb-8730-5f59f1fefac5.png)

But this seems to be an artifact of the model fine tuning since the original model did have a precise mask for this image. However the original model had some flaws with other images the fine tuned model can handle very well. I'll keep you updated and will compare different matting algorithms after model tuning. The current matting model I'm using is this: https://github.com/foamliu/Deep-Image-Matting-PyTorch

**Comment by shivam13juna** [comment]
I see, my algorithms so far aren't using GPU, they're using parallel
processing on CPU. So unless you code CUDA for exploiting GPU resources,
GPU probably wouldn't help. But dividing image has no negative side
effects, and with a good CPU (which was a good CPU on AWS), I extracted a
25-megapixel image (image and mask were input) output in around 10 seconds.
So for my algorithm, a good CPU would probably be a lot helpful. And yeah
man, you gotta fine tune the image mask model, every model, however, seems
to have some problem tho.

**Comment by Nico321** [comment]
Hi, as promised I've also tried your matting algorithm but i have to admit thats a lot worse then the other one i was using. This is what the foreground and background looks like:
![image](https://user-images.githubusercontent.com/6123294/98083560-d565e400-1e7a-11eb-95df-4025616df21b.png)

I've tried different trimaps with similar results:
![image](https://user-images.githubusercontent.com/6123294/98083701-09410980-1e7b-11eb-911c-b0833deb7d07.png)
![image](https://user-images.githubusercontent.com/6123294/98083676-00503800-1e7b-11eb-84f1-2c3105890c74.png)
![image](https://user-images.githubusercontent.com/6123294/98083730-14943500-1e7b-11eb-9327-c5b38719a94b.png)

With the matting algorithm I used so far and the first trimap I get much better results (still not perfect):
![image](https://user-images.githubusercontent.com/6123294/98084083-92f0d700-1e7b-11eb-9e9d-d2fe970cf787.png)
